### ISSUE TITLE
10.11 [Tracking Issue]

### ISSUE THREAD

**Issue by DomT4** [Homebrew/homebrew#40837]
This is essentially a triage of various issues and progress on them. Did the same thing for 10.10 last year and it worked relatively well at keeping everyone in the loop.

This thread will be comment-policed fairly hard; Please keep comments related to failures/issues/etc. If you can reproduce an issue providing a `brew gist-logs FORMULA` is preferred over a simple :+1: - Thanks!

**New issues should still be filed as new issues. This is only a centralised tracking thread rather than a solve-everything-here thread.**

Core Fixes:
* Basic recognition of 10.11 was added to Homebrew - https://github.com/Homebrew/homebrew/commit/62381a65d1aa98e00d3fc2889363d8d44fdf6b58
* `pkg-config` files were added for 10.11 - https://github.com/Homebrew/homebrew/commit/4a8fcd199df144b1d3d209c21f9443d369f4bf5b
* Better detection and warning of unsupported OS X versions was committed - https://github.com/Homebrew/homebrew/commit/baaa295515ce877557dcd69cc89c7e441a6185b2.
* Clang version update for 10.11 DP2 added in https://github.com/Homebrew/homebrew/commit/11cf300c01e3c43fc9604b87072ae12d4208a410.
* Clang version recognition fixed in https://github.com/Homebrew/homebrew/commit/0599646af6e6cf775081afb4fcb9af99d6b1088c.
* Clang version updated for 10.11 DP4 in https://github.com/Homebrew/homebrew/commit/df15afd28b9d7b7778c4f0d84dd8317e5369bacd.

Potential Issues:
* Rootless: Bans writing to [/System, /usr, /bin, /sbin](http://arstechnica.com/apple/2015/06/preview-os-x-el-capitans-first-beta-is-a-promising-heap-of-refinements/4/) - So far `/usr/local` isn't write-protected.
* Permissions can wander off and prevent README overwriting - #41150.
* System OpenSSL: Headers have seemingly been removed from the CLT and Xcode - https://github.com/Homebrew/homebrew/issues/40519#issuecomment-110590553.

Formulae Issues Outstanding:
* Pixman fails to build - #41056 - Can't find somewhere reliable to report upstream.
* Python can fail to build core extension modules - #41085.
* Google-sparsehash fails to build on DP2 - #41087 - [Upstream report](https://code.google.com/p/sparsehash/issues/detail?id=111).
* Fltk fails to build on DP3 - #41581.
* Libgcrypt refuses to build - #41599 - [Upstream report](https://bugs.gnupg.org/gnupg/issue2056).
* Phantomjs refuses to build - #42249 - The internal Qt fails to compile. Needs updating/patching upstream.
* Clang-format - #42221 - You need to symlink `/Applications/Xcode-beta.app/` to `/Applications/Xcode.app/`.
* Ctail refuses to compile - #42400.
* Cputhrottle dies horribly on finding Boost - #42394.
* Ccze fails to compile - #42379.
* Carthage fails to compile, looks like incompatibility with Xcode-Beta Swift - #42378.
* Bup fails to compile - #42375.
* Arabica fails to compile - #42369.
* Postgresql fails to compile - #42223 - Can't find `readline`, needs investigating.
* Libpeas fails to compile - #42685.
* ios-class-guard breaks on trying to find OpenSSL, breaks if you force our OpenSSL - #42671.
* Io fails to compile - #42670.
* Idris fails to compile - #42669.
* Gsoap fails to compile - #42502.

Formulae Fixed:
* Qt fails to build - #40585, #42370 - A patch needs submitting to Qt's Gerrit for review. They aren't supporting it any more. We've backported a patch for 10.11.
* Python & Python3 - #40516 - Tim patched [python](https://github.com/Homebrew/homebrew/commit/80c83ab43a8f1946c301291985c8f2b1fbf073f9) and [python3](https://github.com/Homebrew/homebrew/commit/dfeb73ec5e754a85fc2506b2a16844a6c978e7c3)'s OpenSSL detection.
* Git - #40519 - Misty fixed git to [stop](https://github.com/Homebrew/homebrew/commit/0946193eb69843110fb1f7a2517b74087b10d811) looking for OpenSSL headers when CommonCrypto is used.
* GCC - #40650 - Patch submitted [upstream](https://gcc.gnu.org/bugzilla/show_bug.cgi?id=66523) applied in https://github.com/Homebrew/homebrew/commit/3291699.
* Pango can fail to build against the second Xcode beta - #41086. **FIXED** by the third DP, Xcode 7 Beta.
* Qt5 refuses to build - #42014 - Should be fixed by https://github.com/Homebrew/homebrew/commit/d0020e7a8a6caf558880aef2865c17e4652bba6a.
* MongoDB fails to build - #40955 - Fixed [upstream](https://jira.mongodb.org/browse/SERVER-17728), merged into Homebrew. Should work fine now.
* Rust refuses to build - #41598 - Upstream [report](https://github.com/rust-lang/rust/issues/26292). Fixed upstream, will be fixed here when #42642 is merged.
* Zookeeper fails to compile with `htonll` problems - #42374. Fixed by https://github.com/Homebrew/homebrew/commit/e00142fde31c4873c691e8d22881e68e91e5d343.

**Comment by umhan35** [comment]
:+1:

**Comment by johnwchadwick** [comment]
I can't seem to compile Python anymore, with the new XCode 7/Mac OS X 10.11 CLT. 
```
/System/Library/Frameworks/Foundation.framework/Headers/NSEnumerator.h:7:15: error: expected ';' after @class
@class NSArray<ObjectType>;
```
A [chrome issue](https://code.google.com/p/chromium/issues/detail?id=499448) suggests that it may have to do with some Clang features that haven't been open sourced and I guess therefore not in the CLT release of Clang.

**Comment by DomT4** [comment]
Has there been a new CLT/Xcode update since last week? Can you provide a full `brew gist-logs python`? Thanks! CC @tdsmith

**Comment by johnwchadwick** [comment]
I don't think so. CLT was last updated June 7th.

https://gist.github.com/anonymous/ea54a604214c8da4f10c

Oh, it's worth noting that my doctor issue re: the Python framework probably wasn't there when this log was created. I just installed that via Cask as a stop-gap solution for myself.

**Comment by tdsmith** [comment]
Please open a new issue so we can keep this thread clean; thanks.

**Comment by geoff-codes** [comment]
> Rootless: Bans writing to /System, /usr, /bin, /sbin - So far /usr/local isn't write-protected.

Correct me if I'm wrong, but I believe this _is_ an issue; if you have a `/usr/local` already created at the time you upgrade, you're ok, but with a clean install, one cannot **create** `/usr/local`.

For whatever its worth, I've been using `/Users/Shared/Developer/usr/local` as my prefix for over a year now, and I have not found the old "if you don't use `/usr/local`, stuff will break" adage/warning to be true at all...

**Comment by mikemcquaid** [comment]
> Correct me if I'm wrong, but I believe this is an issue; if you have a /usr/local already created at the time you upgrade, you're ok, but with a clean install, one cannot create /usr/local.

It'll just require a `chown`, no?

> For whatever its worth, I've been using /Users/Shared/Developer/usr/local as my prefix for over a year now, and I have not found the old "if you don't use /usr/local, stuff will break" adage/warning to be true at all.

You're not getting most binary packages which means stuff is much more likely to break.

**Comment by geoff-codes** [comment]
> It'll just require a chown, no?
chown... what? You can't chown `/usr`, and with `sudo mkdir /usr/local` acts as though one's trying to to write to a read-only volume (which is, I believe, basically how "rootless" is implemented).

> You're not getting most binary packages which means stuff is much more likely to break.
If you mean I'm not getting the bottled binary packages, yes, those would break; but back in February, PRs #37059 - #37072 were the result of an experiment I did, attempting to install every homebrew formula in this directory. Essentially, there were the only ones that failed (for unrelated reasons). I may not have tested each package thoroughly, but I'd say (generally) it was a success.

**Comment by mikemcquaid** [comment]
This doesn't match what I've seen; I've been installing Homebrew fine onto a clean 10.11 VM into `/usr/local`.

**Comment by geoff-codes** [comment]
> This doesn't match what I've seen; I've been installing Homebrew fine onto a clean 10.11 VM into /usr/local.

I'll double check.

**Comment by geoff-codes** [comment]
> This doesn't match what I've seen; I've been installing Homebrew fine onto a clean 10.11 VM into /usr/local.

@mikemcquaid Ok you're right, but I'm still (kind of) right-ish:

1. If you **upgrade** a machine which has an existing `/usr/local`, the permissions are maintained, and special xattrs are applied to the directory to keep it writable;
2. I you do a completely clean install, _Apple creates_ a `/usr/local` with these special xattrs, and you can write to this directory (they haven't done that in a while!);
3. If you upgrade a machine which **does not** have an existing `/usr/local` when you upgrade, Apple **does not** create this directory;
4. If the `/usr/local` with special xattrs is deleted, one cannot re-create it.

So... still a bit of an issue, I'd say?

**Comment by mikemcquaid** [comment]
I wouldn't necessarily rely on this behaviour remaining consistent to the actual release, based on past experience. I believe that we should still be able to create that directory somehow.

**Comment by geoff-codes** [comment]
> I wouldn't necessarily rely on this behaviour remaining consistent to the actual release, based on past experience. I believe that we should still be able to create that directory somehow.

I agree. If there isn't a way to re-create in in later betas, its probably time to raise a stink. I believe FHS and POSIX require `/usr/local` to writable, and I don't think Apple would want to risk losing SUS UNIX certification over this.

**Comment by mikemcquaid** [comment]
Are you sure it's literally impossible for a `sudo`d user to create those directories with the `xattrs`?

**Comment by geoff-codes** [comment]
Quite — at least by any conventional, reasonable method I know of.

```
Tests-MacBook-Pro:~ test$ rm -rf /usr/local
Tests-MacBook-Pro:~ test$ mkdir /usr/local
mkdir: /usr/local: Operation not permitted
Tests-MacBook-Pro:~ test$ sudo mkdir -p /usr/local
mkdir: /usr/local: Operation not permitted
Tests-MacBook-Pro:~ test$ ruby -e "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/master/install)"
==> This script will install:
/usr/local/bin/brew
/usr/local/Library/...
/usr/local/share/man/man1/brew.1

Press RETURN to continue or any other key to abort
==> /usr/bin/sudo /bin/mkdir /usr/local
mkdir: /usr/local: Operation not permitted
Failed during: /usr/bin/sudo /bin/mkdir /usr/local
```

The only feasible method I can think of is to do something like:
- Gain sudoer privileges
- `sudo nvram boot-args="rootless=0"
- Write a LaunchAgent, and restart. The LauchAgent will:
   - create `/usr/local` and set the proper permissions
   - resent `nvram boot-args`
   - delete itself
   - restart again

But this method is also the basic rubric for installing nigh-undeleteable malware.

Also the following files exist:
```
/System/Library/Sandbox/rootless.compat
/System/Library/Sandbox/rootless.conf
/usr/libexec/rootless-init
```

But these are completely undocumented and how they work is anyone's guess, I think.

(My personal opinion: I understand why Apple implemented this for iOS, but rootless for OS X is a pretty half-baked idea. As I mention above the fact that one can turn it off and on at will makes it a great way to install malware which effectively cannot be removed. Myself, I see `rootless=0` becoming standard across my deployments if its released in its present form.)

**Comment by geoff-codes** [comment]
As regards re-creating `/usr/local` with rootless enabled, this persists with beta 2; I've taken the liberty of opening rdar://21542005 at bugreport.apple.com.

Contents (In case someone wants to recreate on OpenRadar or elsewhere):
 
> 10.11: Rootless — Inability to re-create /usr/local if deleted.

> Summary:
With the default configuration (rootless=1), it is possible for the root user to delete or move `/usr/local`, but it is *not* possible for the root user to re-create `/usr/local`. Per the WWDC '15 presentation "Security and Your Apps", `/usr/local` is to be "available to developers" (http://f.cl.ly/items/362w042M1V2C2b273313/Screen%20Shot%202015-06-12%20at%2012.57.46%20AM.png); this is not necessarily the case presently.

> This is a particularly impediment to homebrew (which I imagine is largely why access to this directory was promised in the first place); please see:
> https://github.com/Homebrew/homebrew/issues/40837#issuecomment-113428804 and linked issues.

> If my understanding is correct, 10.11 SIP only allows an installer package signed by the Apple Software Update CA (and other Apple CAs [?]) to modify a "rootless" protected directory. Therefore, presently there appears to be no means to re-create the directory without reinstalling Essentials.pkg in its entirety. It would be great if another means were added to rectify such a situation, such as:

> - diskutil repairSomething /
- A properly signed package (contents: an empty /usr/local) were kept somewhere on the system
- Add some magic to `mkdir`.
- ?

> Steps to Reproduce:
1. `sudo rm -rf /usr/local`.
2. `sudo mkdir /usr/local`.


> Expected Results:
Ambiguous (?)

>Actual Results:
> `mkdir: /usr/local: Operation not permitted`
No (apparent) means to recreate without a complete reinstall.

> Version:
10.11 beta 1; 10.11 beta 2

> Notes:
I'm filing this under "Classification: Security" because this is a bug caused by a Security framework, not because it is a "security issue". I would still, respectfully, consider this a "serious bug".

> Configuration:
Occurs on 10.11 beta 1 or beta 2.
Does not occur with "rootless=0" or on previous systems.


I'll report back if I get any updates.

**Comment by mikemcquaid** [comment]
Thanks @geoff-codes. I wouldn't worry too much; I'd be very surprised if the current behaviour survives to GA.

**Comment by geoff-codes** [comment]
> I wouldn't worry too much; I'd be very surprised if the current behaviour survives to GA.

Yeah I've just been trying out more of an "if you see something, say something" approach these days :inbox_tray:.
Can't hurt, anyway. :wink:

**Comment by mikemcquaid** [comment]
> Yeah I've just been trying out more of an "if you see something, say something" approach these days .
> Can't hurt, anyway.

Definitely. We always appreciate upstream bug reports about such things!

**Comment by DomT4** [comment]
Everything with a `*` next to it has write access with `sudo`. Everything on the list without that star is blocked to everyone:

```
				/System
*				/System/Library/Caches
booter			/System/Library/CoreServices
				/System/Library/CoreServices/RawCamera.bundle
*				/System/Library/Extensions
				/System/Library/Extensions/*
UpdateSettings	/System/Library/LaunchDaemons/com.apple.UpdateSettings.plist
*				/System/Library/Speech
*				/System/Library/User Template
				/bin
dyld			/private/var/db/dyld
				/sbin
				/usr
*				/usr/libexec/cups
*				/usr/local
*				/usr/share/man
# symlinks
				/etc
				/tmp
				/var
```

I'm not quite sure why `/tmp` is on the list; I can seem to write to that without sudo just fine. `/etc` is protected and can't write without `sudo`, as is `/var`.

There's also a small exemption list that permits certain non-Apple binaries in `/usr/bin` and `/usr/sbin`.

**Comment by mikemcquaid** [comment]
Most things seem to find OpenSSL on 10.11 as long as it's installed into `/usr/local`. If not, it's a bit of a shitshow and almost everything seems to fail to find it.

**Comment by DomT4** [comment]
I was having a serious ponder about unkegging OpenSSL, but I'm not sure how much hell that would cause.

**Comment by mikemcquaid** [comment]
On 10.11 that seems like a very good idea (and matches what we've done with other software). I can see an argument for doing it for all OS X versions at this point.

**Comment by DomT4** [comment]
My other thought was permanently adding our OpenSSL paths to superenv. Given the amount of stuff that ties into these days it's extremely likely someone would have it installed, or install it fairly quickly, but if felt like a hack so I went off the idea.

Interested in what @mistydemeo thinks on unkegging, as it's possible some of the older OS X versions could have stuff that dynamically links into a system OpenSSL, but if you can no longer compile many things against a system OpenSSL on 10.11 it would seem reasonable for Homebrew to fill that void.

An alternative is finding everything that has an OpenSSL link in Homebrew between now and 10.11 release and ensuring it can find and use Homebrew's OpenSSL, and adding revisions to those formulae.

**Comment by mikemcquaid** [comment]
I think just not making it keg-only is probably the way forward. Given the security situation I think we can do that globally unless any @Homebrew/owners have objections.

**Comment by DomT4** [comment]
We have an OpenSSL bump landing on the 9th, and I'm planning to try and handle the SSLv3 changes shortly after, so we could unkeg it either with the bump, or with the revision-all-the-things PR.

**Comment by mikemcquaid** [comment]
@DomT4 With the bump sounds good.

**Comment by chdiza** [comment]
If `/etc` can't be modified at all, how is one supposed to append a custom shell to `/etc/shells`?  AFAIK  the `bash` and `zsh` formulas will be dead without this ability.

**Comment by zmwangx** [comment]
@chdiza Without a working system to test, I would guess that

```
# symlinks
                /etc
                /tmp
                /var
```

in https://github.com/Homebrew/homebrew/issues/40837#issuecomment-115528215 really means the symlinks themselves can't be modified. The directories `/private/etc`, `/private/tmp` and `/private/var` should remain writable (it would be insane otherwise, especially for `/private/tmp`).

**Comment by DomT4** [comment]
Yes. You can still seem to write files to `/etc`, `/tmp` and `/var`. It hasn't been explained what Apple means by `symlinks` in that document.

**Comment by galxy25** [comment]
Anyone been able to do a clean install of node on 10.11? I tried and got lgcc libraries missing errors which prevented the install.

**Comment by zmwangx** [comment]
@galxy25 Yes, I was able to compile 0.12.7 without any problems.

**Comment by galxy25** [comment]
Thanks, I found a closed issue on the home brew repo related to this. Boiled down to me needing to use the build tools for Xcode 7 beta versus 6.4

**Comment by chriskrycho** [comment]
FWIW, `sudo chown -R $USER /usr/local` made brew work properly again for me.

**Comment by smason** [comment]
I also deleted `/usr/local` during an upgrade and have been struggling with trying to get everything working…  It's taken a few attempts to figure out that the *correct permissions* mentioned above involve removing the `restricted` flag on `/usr/local`, i.e:

    sudo chflags norestricted /usr/local

after setting the `rootless` boot-flag first—verify with `ls -lO /usr`.

Can't find many mentions to this now I know what to search for, but any further comments would be appreciated!

**Comment by geoff-codes** [comment]
@smason
Yep, you've got the right approach, for now.

The radar bug I filed on has been updated with the note:
`Duplicate of 20967238 (Open)`

Now of course, I can't see what rdar://20967238 actually is, but it would appear that as @mikemcquaid correctly predicted, Apple is already aware of the issue, and it very likely will be fixed before GM.

**Comment by apjanke** [comment]
FYI, the system `zsh` behavior has changed slightly in 10.11. In the public beta, the `$FPATH` for the system `zsh` now contains `/usr/local/share/zsh/site-functions` by default. So anything Homebrew formulas install there will be visible to the system `zsh` by default, instead of requiring it to be added in the startup files. This appears to be compiled in, and not part of the global configuration in `/etc/z*`.

This sounds like it's related to the "rootless" stuff locking down `/usr/share` so programs couldn't install completion functions there any more.

**Comment by apjanke** [comment]
I did bfontaine's schtick and attempted to `brew install --build-from-source` all formulae in core Homebrew on an OS X 10.11 Public Beta 2 test box. (Relatively clean install, no CLT.) Here's the list of formulae for which it failed: https://gist.github.com/apjanke/45b50d2ae540e396dba0.

I'm currently re-running the failed installs with `HOMEBREW_MAKE_JOBS=1` to produce debugging log output and will post individual issues with logs as I get a chance. Since I did the initial install attempts, Public Beta 3 came out and some of the re-runs are succeeding, but I don't know yet if that's because of PB3 or running it single-threaded, or side effects of having installed other formulae.

**Comment by DomT4** [comment]
Are you sure `qt5` doesn't compile? I added the upstream patch for 10.11 (Or rather, Xcode-Beta's Clang) a couple days back which should fix that.

**Comment by apjanke** [comment]
I'm not sure about any of these failures, really: the compile job took about five days to run so it probably didn't have the `qt5` patch, and Public Beta 3 came out after that. The rerun of the failed jobs should pick that up.

**Comment by apjanke** [comment]
`qt5` still fails for me, even with the patch. I've updated #42014 with my logs.

**Comment by apjanke** [comment]
I revised the script to only note formulae which failed themselves, not ones whose dependencies are failing.

Here's an updated list.

https://gist.github.com/apjanke/45b50d2ae540e396dba0#file-failed-public-beta-3-no-clt-excluding-dependent-failures-txt

**Comment by apjanke** [comment]
MongoDB may be fixed, but libmongoclient is still failing. 10.11 isn't supported by their build system. Reported upstream: https://jira.mongodb.org/browse/CXX-648

**Comment by apjanke** [comment]
FYI, I've switched to tracking my test installation failures on a Google Docs spreadsheet, [Homebrew El CapitanFailures](https://docs.google.com/spreadsheets/d/1a1PsKkscxYUCV4gLL5lvnVb7PeJvdJTr7VAvdo1FEhA/edit?usp=sharing). Should be world-readable. I'll continue noting upstream bugs in batches here.

The harness code I used to run these tests (including the basic results data) is in the [apjanke/brew-el_cap-work repo](https://github.com/apjanke/brew-el_cap-work). (This has probably been done better elsewhere.)

Homebrew installation logs are generally not a security or privacy concern, right? I could check in a tarball of my logs from the test runs if people are interested in the results and don't wan to spend the week or so of execution time to run everything themselves.

**Comment by mikemcquaid** [comment]
> Homebrew installation logs are generally not a security or privacy concern, right? I could check in a tarball of my logs from the test runs if people are interested in the results and don't wan to spend the week or so of execution time to run everything themselves.

Yeh, they are generally not a privacy concern unless you're worried about your e.g. username.

**Comment by zmwangx** [comment]
@apjanke Thanks for building all core formulas, but here's a question: did you run `brew test` on each successfully built formula? I didn't read your scripts, but a [quick search](https://github.com/apjanke/brew-el_cap-work/search?utf8=%E2%9C%93&q=test) doesn't reveal any `brew test` invocation.

The reason I'm asking is because I just noticed a formula that had no problem building but fails when I actually use it. `brew test` was able to capture the failure. While that formula was eventually fixed by reinstalling, running `brew test` on formulas is certainly a good way to catch additional problems.

**Comment by apjanke** [comment]
Nope, I didn't. That's a good idea. I'll add it to the harness code. Probably `brew audit`, too.

**Comment by apjanke** [comment]
Upstream issue for mongrel2: mongrel2/mongrel2#279

**Comment by apjanke** [comment]
Any chance of getting an el cap bottle built for `boost`? It's used by a lot of formulae, and takes a long time to build.

```
$ brew uses boost | wc -l
     159
```

**Comment by apjanke** [comment]
Other failures:

- [ ] pincaster - can't find OpenSSL - #43991
- [ ] pkcs11-helper - can't find OpenSSL - #43992
- [ ] racket - build failure - #43993 [upstream bug here](http://bugs.racket-lang.org/query/?cmd=view&pr=15154)

**Comment by apjanke** [comment]
- [ ] proxytunnel - Can't find OpenSSL - #43995
- [ ] sdhash - Can't find OpenSSL - #43997
- [ ] skipfish - Can't find OpenSSL - #43999
- [ ] snownews - Can't find OpenSSL - #44001
- [ ] sofia-sip - Can't find OpenSSL - #44002
- [ ] sourcekitten - Dependency on specific Xcode version - #43684
- [ ] spatialite-gui - Not linking to `sqlite3` lib - #44003

**Comment by apjanke** [comment]
- [ ] sstp-client can't find openssl, even though it `depends_on "openssl"` - #44008
- [ ] stuntman - Can't find OpenSSL - #44009
- [ ] swiftlint - Swift version incompatibility - #44010
- [ ] tcpreplay - Can't find libpcap inside Xcode - #44011
- [ ] telepathy-gabble - Compiler error w redefined function prototype - #44014
- [ ] tls-date - Can't find OpenSSL - #44015
- [ ] tn5250 - Can't find OpenSSL - #44016
- [ ] valgrind - Need to recheck against hardcoded paths, 32/64 - #38555

**Comment by mikemcquaid** [comment]
> Any chance of getting an el cap bottle built for boost? It's used by a lot of formulae, and takes a long time to build.

I've tried a few times and although it builds locally it seems to hang indefinitely building on the bot, annoyingly.

**Comment by apjanke** [comment]
More failures:
- [ ] vault - `go` compilation error - #44035
- [ ] volatility - fails to find freetype headers - #44039
- [ ] voms - Can't find OpenSSL - #44040
- [ ] whereami - Swift compilation failure - #44044
- [ ] xaric - Can't find OpenSSL, even though it `depends_on "openssl"` - #44048
- [ ] xsp - Can't find `gmcs` - #44051

**Comment by apjanke** [comment]
- [ ] cabal-install - Illegal instruction in bottled binary - #43985

**Comment by apjanke** [comment]
- [ ] nsd - Can't find openssl, even with `depends_on "openssl"` - #44056
- [ ] openal-soft - Build failure - #44059
- [ ] openh264 - Build failure - #44060

**Comment by apjanke** [comment]
Okay, I've worked through the whole list of the build failures from my "Try installing everything" run. All the breakage I found has been turned in to bug reports, PRs, or items on this issue. The [Google Sheet](https://docs.google.com/spreadsheets/d/1a1PsKkscxYUCV4gLL5lvnVb7PeJvdJTr7VAvdo1FEhA/edit#gid=0) is updated to reflect it.

Let me know if I can do anything else to help out here.

**Comment by RandomDSdevel** [comment]
Just wanted to note that Apple quietly updated its El Cap web page with a release date of September 30<sup>th</sup> after the 'Hey, Siri!' event on the 9<sup>th</sup>.

**Comment by ablyler** [comment]
@mikemcquaid i tried to debug why brew bot was having such a rough time with boost, and it looks like it is the verbose flag.  removing it caused the build to finish in a few minutes on my machine, with the verbose flag it ran for hours and didn't finish.

**Comment by ablyler** [comment]
@mikemcquaid I created a bottle of boost locally on my OS X 10.11 box. I've uploaded it to: https://cudadrive.com/djQgcQYpSu6gDgH2

The bottle block is:
```
  bottle do
    cellar :any
    revision 1
    sha256 "556e6c649ecfc21c40a06b42c8742bccde73676c35b3ebb40487e969faad6f3e" => :el_capitan
  end
```

**Comment by DomT4** [comment]
I actually got an email on the Boost situation yesterday afternoon:

> I updated my Xcode to the last version 7. It breaks any C++ code using the BOOST_STATIC_ASSERT macro. I filled a bug report for boost, #11664, URL https://svn.boost.org/trac/boost/ticket/11664

> I don't know if/when it will be fixed in a boost release, anyway the ticket proposes a (tested) fix.

> As El Capitan should come soon (and the corresponding Xcode is already available and pushed by Apple) I strongly suggest to schedule a boost formula update for as soon as possible.

> Thanks

> Francis Dupont <fdupont@isc.org>

**Comment by zmwangx** [comment]
Might or might not be related to "Permissions can wander off and prevent README overwriting" #41150, but I just provisioned a clean GM install, and with a non-interactive, non-sudo (run from my admin account with temporary `%admin  ALL=(ALL) NOPASSWD:ALL` in `/etc/sudoers`)

```zsh
ruby -e "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/master/install)" </dev/null
```

I somehow ended up with a `/usr/local` with `root:admin` (I think it used to be `$USER:admin`, but it seems that now it's only made group-writable?). Moreover, `/usr/local/.git/FETCH_HEAD`, `/usr/local/.git/ORIG_HEAD`, and the `.git/FETCH_HEAD`, `.git/ORIG_HEAD` of all tap repos are all `root:admin` with mode `644`, so that `brew update` can't be performed since all `FETCH_HEAD` and `ORIG_HEAD` do not belong to the admin user and are not group-writable. I had to `chown -R $USER:admin` on all `.git` directories to make it work.

I'll investigate when I have time, but has anyone else encountered this problem?

**Comment by DomT4** [comment]
There's now a document for this [here](https://github.com/Homebrew/homebrew/blob/master/share/doc/homebrew/El_Capitan_and_Homebrew.md) which the Troubleshooting page points at. Hopefully clears up future confusion/problems/etc. We'll be adding a `brew doctor`/etc warning soon as well.

**Comment by zmwangx** [comment]
> Hopefully clears up future confusion/problems/etc.

Yeah, I know how to fix the issue, but shouldn't we make the install script do the right thing in the first place? (Again, haven't investigated yet.)

**Comment by Echelon9** [comment]
Noting on this meta issue tracker https://github.com/Homebrew/homebrew/issues/44291 (Upgrade Valgrind to 3.11.0).

**Comment by routelastresort** [comment]
FWIW, I had to follow the second part of the [permissions doc](https://github.com/Homebrew/homebrew/blob/master/share/doc/homebrew/El_Capitan_and_Homebrew.md) even though /usr/local already existed, as `/usr/local/bin` (and maybe `lib` at one point) kept reverting to root ownership.  This wasn't happening before, so the behavior might have changed around the gold release.  It looks like `chflags norestricted` and re-enabling rootless was the only way to stop things from happening behind my back.

**Comment by fjarlq** [comment]
Damn. I just got bit by this while upgrading to El Capitan. Now I can't recreate my `/usr/local` directory. Here's what happened:

1. Remembering that the OS X upgrade mysteriously takes hours to process my Homebrew `/usr/local`, I moved `/usr/local` into my homedir prior to starting the El Capitan upgrade.
2. I created an empty `/usr/local` just before starting the upgrade, thinking something in the installer might break if it doesn't find one.
3. The upgrade went fine; took about an hour I suppose.
4. Back up on El Capitan, I see that the empty `/usr/local` is there, owned my `root:wheel` and mode `0755`
5. So, I `sudo rmdir /usr/local` in preparation to `mv` the Homebrew `/usr/local` back into place.
6. `mv` fails with `Operation not permitted` and I discover this whole mess.

Has anybody figured out a good way to recreate `/usr/local` in this situation?

**Comment by DomT4** [comment]
@fjarlq https://github.com/Homebrew/homebrew/blob/master/share/doc/homebrew/El_Capitan_and_Homebrew.md#if-usrlocal-does-not-exist

Theoretically in your case, instead of creating `/usr/local` you could just move the old one back after SIP is disabled and then follow the rest of the command.

FWIW, I haven't seen any reports of 1 happening yet this year. Not saying it won't to anyone, but *possibly* less of an issue.

**Comment by fjarlq** [comment]
Thanks, @DomT4. A variant of that worked for me. Here's what I did:

1. Rebooted into Recovery Mode and brought up Terminal.
2. `cd /Volumes/myroot`
3. `mv Users/myuser/local usr`
4. `chflags norestricted usr/local`
5. Rebooted back into regular mode.

**Comment by pboling** [comment]
The hours long upgrade issue did not happen to me just now.  Upgraded to El Capitan in under an hour all told (I did have to re-install Xcode, apparently El Capitan wipes the old install).

**Comment by lifepillar** [comment]
I assume https://github.com/Homebrew/homebrew/blob/master/share/doc/homebrew/El_Capitan_and_Homebrew.md#if-usrlocal-does-not-exist is the current official way of installing Homebrew on El Capitan.

My 2 cents: that is sailing against the wind. While that is possible, in my experience it is seldom a good idea in the long term. Maybe, it is time for Homebrew's developers to review some design choices. Which in this case means one of two possibilities:

1. living with the default permissions of `/usr/local` and accept that `sudo` is necessary to install stuff outside of a user's home directory;

2. totally embrace the "no-sudo" approach, and change Homebrew's location to a place inside a user's home directory (with all the consequences).

I love Homebrew, and installing packages without `sudo` has probably been one its main selling points. And I know that Homebrew was designed having in mind a single-user development environment and not a multi-user production environment. But, thinking in terms of convenience, what is more convenient now for the user? Rebooting a system twice from time to time, or requiring a password for `brew install` and `brew upgrade` (like every other package manager in the world)? Apart from convenience, what other arguments are there in favor of a "no-sudo" Homebrew? And isn't Homebrew mature enough to be a package manager for production environments?

How about having Homebrew installed in a user's home directory, but the Cellar in `/usr/local`?

**Comment by zmwangx** [comment]
> living with the default permissions of `/usr/local` and accept that `sudo` is necessary to install stuff outside of a user's home directory;

Moot point. Homebrew trades `sudo`ing over and over for `sudo`ing once up front, during installation. It would be incredibly stupid if `brew upgrade` requires root permissions (after all these years), since I daresay a big chunk of users are constantly on the bleeding edge and upgrade daily, probably through automated jobs.

> Rebooting a system twice from time to time.

Define "from time to time". You mean, after each system upgrade? Actually I don't understand "kept reverting to root ownership" mentioned above. The claim is not clear at all. Define "kept". Does that mean "after reboot"? Or "after toggling SIP"? At least I haven't seen that myself after a clean install of the final released build. Moreover, with SIP I can chmod on `/usr/local` without problem; no reboot required, let alone twice. Maybe I just haven't lived with it long enough, but I honestly don't think what hasn't happened today will magically happen tomorrow.

> And isn't Homebrew mature enough to be a package manager for production environments?

No one stops you from running Homebrew from the root account "for production", as far as I could tell. The ownership will be `root:admin` rather than `root:wheel`, but I don't see how that's a big problem. Most things are not group writable anyway. The only caveat is that you have to keep using the root user to manage everything; not very different from `apt`.

> How about having Homebrew installed in a user's home directory, but the Cellar in `/usr/local`?

Ridiculous.

**Comment by lifepillar** [comment]
>a big chunk of users are constantly on the bleeding edge and upgrade daily, probably through automated jobs.

Such users are likely clever enough to fix their automated jobs once and for all. It surely would be less of a hassle than having such jobs periodically fail because permissions are reset.

>Define "from time to time". You mean, after each system upgrade?

According to the link above, yes. But OS X 10.11 might check permissions at other times as well (a hint is that the Repair Permissions function in Disk Utility is not needed any longer). However, `/usr/local` may not be affected.

>Moreover, with SIP I can chmod on /usr/local without problem; no reboot required, let alone twice.

Ah, a clean install already has `/usr/local` and you may change permissions on it. So, Recovery Mode is only needed if, for some reason, such directory does not exist. The question remains, whether such permissions are reverted by OS X or not, and when.

>No one stops you from running Homebrew from the root account "for production", as far as I could tell. 

As long as the feature is not officially supported (last time I checked it wasn't), no thanks.

>>How about having Homebrew installed in a user's home directory, but the Cellar in /usr/local?
>Ridiculous.

Reasonable, if Homebrew is owned by me and the formulas are installed with administrator privileges. Anyway, If OS X lets you change the rights on `/usr/local` without interfering, I have no problem in keeping things as they are. But if the system reverts the changes I make then the message is clear: this is not the way you should do it, like it or not.

**Comment by DomT4** [comment]
Some replies which hopefully help handle various points raised before this discussion takes over the thread:

* At this time we aren't considering moving to another default prefix. `/usr/local` remains accessible to developers per Apple's own Keynote & everyone who writes into `/usr/local` and don't root all the things is going to deal with the same issues we are.

* I'm personally very strongly against reversing our `sudo` issue. You're either going to be using `sudo` for *every* brew command that writes into the prefix, or you use `sudo` to fix permissions every now and again. There's also the previously made argument of arbitrary code installing with full root access.

* What the situation is with SIP is unclear going forwards. Apple's public documentation and Keynote points on `/usr/local` are *vague*. It is implied that the permissions will be reset on updates (From 10.11 to 10.11.1 for example) but I haven't seen evidence of this yet. Likewise with `Disk Utility` resetting permissions on the directory. It's too early into the SIP timeline to make drastic decisions here.

* There are some 2231 formulae with bottles, 1566 of which are relocatable. Moving prefix would involve rebuilding 665+ binary packages which could take weeks if nobody submits a PR & we don't do any core changes in that time, and a lot longer if anyone does.

* Every other tap in the Homebrew ecosystem would have to also migrate their bottles over if those bottles aren't relocatable.

* We could migrate people's data over to the new prefix but things would still break. Some things hardcode their prefix and moving that causes failure.

* A lot of automated scripts elsewhere would break. Bear in mind that Travis' OS X machines ship with Homebrew installed and using Homebrew for testing projects elsewhere is very common.

People can do what they like with their Homebrew installations, but we aren't going to rush a decision here and the plan for the foreseeable is to stay in `/usr/local` by default.

**Comment by zmwangx** [comment]
@lifepillar 
To avoid taking over the thread, I won't respond to every point raised, but just want to point out an obvious flaw in your argument: at one point you are worried about having to boot into recovery "from time to time", but then you answered yourself:

> Ah, a clean install already has /usr/local and you may change permissions on it. So, Recovery Mode is only needed if, for some reason, such directory does not exist. The question remains, whether such permissions are reverted by OS X or not, and when.

Okay, I'll just assume for now that OS X reverts /usr/local's permissions without notice from time to time. But what won't ever happen is /usr/local being deleted once it's already there (without action on the user's part anyway). So where does that booting into recovery from time to time coming from?

**Comment by lifepillar** [comment]
@zmwangx I have realized after your reply that Recovery Mode is necessary only under certain circumstances.

**Comment by Mikaela** [comment]
> living with the default permissions of /usr/local and accept that sudo is necessary to install stuff outside of a user's home directory;

:+1:. I am breaking the guidelines by doing this, because I have multi-user (family) systems and it's possible for someone else to need to install something. Currently Homebrew is assuming all systems are only single user which can cause trouble when it's not true.

**Comment by DomT4** [comment]
`brew` should actually completely happily run under `sudo` if the executable is owned by root. We don't actively support the configuration, but we don't block it in the same way we do if the executable isn't owned by root.

**Comment by MStines007** [comment]
^^^ `sudo brew update` worked without issue, for me.

**Comment by cwonrails** [comment]
Speaking as someone who's been upgrading with every 10.11 beta and struggled endlessly with SIP, :+1: for @DomT4's comments [above](https://github.com/Homebrew/homebrew/issues/40837#issuecomment-144782245). For those seeking change, his conclusion warrants special attention:

> People can do what they like with their Homebrew installations, but we aren't going to rush a decision here and the plan for the foreseeable is to stay in /usr/local by default.

As others have noted and many have experienced, Apple's implementation of and communication on SIP has been (at best) inconsistent. I can confirm that in my local environment (latest public beta) the permissions on `/usr/local` consistently reset of their own accord even when the user disables SIP and `chown`s the relevant directories. That does _not_, however, mean Homebrew should change default behaviors and environmental norms that users have internalized and relied upon for years. The onus is on Apple to align 10.11's real-world treatment of `usr/local` with the information presented at WWDC.

**Comment by ivan-leschinsky** [comment]
I found the answer to disable new security layer in the OS X 10.11.
In the recovery mode(reboot with Cmd+R) in the console:
```sh
csrutil disable
```
From here(last answer): https://apple.stackexchange.com/questions/101328/file-cant-be-moved-by-root-on-os-x/202969#202969?newreg=0b6a3e5858b5401bbe9e3040f9085f4c

**Comment by DomT4** [comment]
@ivan-leschinsky Please be aware Homebrew isn't recommending doing that for more than the length of time it takes to fix whatever permission issues you are seeing, unless you have other reasons to want full root access. We believe rootless is generally a good thing for the vast majority of users.

**Comment by ivan-leschinsky** [comment]
@DomT4 Got it. But disabling rootless doesn't helped me.

**Comment by DomT4** [comment]
@ivan-leschinsky What problems are you seeing?

**Comment by ivan-leschinsky** [comment]
@DomT4 I saw `Operation not permitted` error. But all is well now! Thanks :smile:

**Comment by DomT4** [comment]
@ivan-leschinsky Glad to hear! Let us know if you hit any further issues :)

**Comment by apjanke** [comment]
Looks like there's a mongoclient update; see #44724.

**Comment by magic890** [comment]
Have you solved this problem?
I'm using `Homebrew 0.9.5 (git revision 51a477; last commit 2015-08-03)` and I'm still getting this error:

    brew update
    Error: The /usr/local directory is not writable.
    Even if this directory was writable when you installed Homebrew, other
    software may change permissions on this directory. Some versions of the
    "InstantOn" component of Airfoil are known to do this.

    You should probably change the ownership and permissions of /usr/local
    back to your user account.
      sudo chown -R $(whoami):admin /usr/local

Running `brew upgrade` I was able to update some formulas.

**Comment by bfontaine** [comment]
@magic890 did you try running the suggested command?

**Comment by wspurgin** [comment]
> System OpenSSL: Headers have seemingly been removed from the CLT and Xcode - #40519 (comment).

Haven't hit that issue just yet, but I ran [across this interesting article](https://langui.sh/2015/07/24/osx-clang-include-lib-search-paths/) that explains *why* they were removed

**Comment by magic890** [comment]
@bfontaine I ran `brew doctor` and tried to use `sudo`. Is that right?

**Comment by wspurgin** [comment]
@magic890 I think @bfontaine means have you used this command:
```sudo chown -R $(whoami):admin /usr/local```

I ran across the same issue you had, but the above command solved it for me.

**Comment by magic890** [comment]
Ok, now it is fixed. Thank you.

**Comment by MatthewHannigan** [comment]
You should use the *-h* option with chown, because you have to be careful in case the link points outside of /usr/local.  And of course the link itself would not get changed, which might inhibit future homebrew operations.

I had a look at my /usr/local and there was a single link which linked outside;
    /usr/local/Cellar/redo/0.11/lib/redo/sh -> /bin/sh

With El Capitan, chown'ing this results in `Operation not permitted`

In general, with almost any use of chown you want the *-h* flag.  I would call it's normal behaviour a mis-design.

**Comment by DomT4** [comment]
Closing as most of the issues here are dealt with and fixed. The remaining ones are:

* https://github.com/Homebrew/homebrew/issues/44051 - Unrelated to OS X 10.11 specifically.
* https://github.com/Homebrew/homebrew/issues/44060 - Is a `nasm` bug.

Thanks for coming along for the ride all. 10.11 was less a nightmare than 10.10, so that's something :smile_cat:.

**Comment by mikemcquaid** [comment]
Thanks for the great work here everyone (particularly @apjanke and @DomT4!)

**Comment by apjanke** [comment]
You're welcome! Glad this went well.

**Comment by lloeki** [comment]
The behaviour once `/usr/local` is deleted is still the same on 10.11.4 with rootless enabled, so for those landing here for that reason, the easiest way is to boot into recovery, unlock and mount the `Macintosh HD` partition via Disk Utility (if encrypted), and start a terminal:

```
$ mkdir "/Volumes/Macintosh HD/usr/local"
$ chflags norestricted "/Volumes/Macintosh HD/usr/local"
$ ls -lOd "/Volumes/Macintosh HD/usr/local"
drwxr-xr-x  2 root  wheel  - 68 17 Mar 09:24 /Volumes/Macintosh HD/usr/local
```

PS: issue participants, sorry for the noise, but the pieces are hard enough to gather and this is quite a showstopper when it happens.

**Comment by lloeki** [comment]
@DomT4 @mikemcquaid Since the solution to a `/usr/local` disappearance isn't automatable and I'm not sure this warrants a Homebrew doc/wiki update, I made a [Stack Exchange answer](http://apple.stackexchange.com/questions/234307/how-to-recreate-usr-local-with-sip-rootless-mode-en-el-capitan/234308).

**Comment by ilovezfs** [comment]
No, it is no longer the case that disabling SIP is required in order to recreate /usr/local, sorry, and it has not been true for quite some time.

See the following:
https://github.com/Homebrew/legacy-homebrew/issues/45387
https://github.com/Homebrew/legacy-homebrew/pull/45566

Stack Exchange answers telling users to disable SIP and giving step by step instructions on how are highly irresponsible, especially when they are entirely unnecessary, and I'd really appreciate it if you would take it down.
